### PR TITLE
rocAL - External source reader

### PR DESCRIPTION
**Pull request by SundarRajan98**
Added 3 Modes for external source reader whose sources varies accordingly :
• Mode 0 - File Names
• Mode 1 - Raw Buffer
• Mode 2 - Decoded Buffer
• Added C++ and python examples for all 3 modes

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -33,7 +33,14 @@ find_program(MAKE_NSIS_EXE makensis)
 find_program(RPMBUILD_EXE rpmbuild)
 find_program(DPKG_EXE dpkg)
 
-set(ROCM_PATH /opt/rocm CACHE PATH "Default ROCm installation path")
+# ROCM Path
```
Comment (rrawther) [inline CMakeLists.txt:36]: I think this is not needed anymore.
@kiritigowda: can you confirm

File: rocAL/include/meta_data/meta_data_reader.h [hunk 2]
```diff
@@ -60,7 +60,7 @@ struct MetaDataConfig {
     bool _aspect_ratio_grouping;
 
    public:
-    MetaDataConfig(const MetaDataType& type, const MetaDataReaderType& reader_type, const std::string& path, const std::map<std::string, std::string>& feature_key_map = std::map<std::string, std::string>(), const std::string file_prefix = std::string(), const unsigned& sequence_length = 3, const unsigned& frame_step = 3, const unsigned& frame_stride = 1)
+    MetaDataConfig(const MetaDataType& type, const MetaDataReaderType& reader_type, const std::string& path = std::string(), const std::map<std::string, std::string>& feature_key_map = std::map<std::string, std::string>(), const std::string file_prefix = std::string(), const unsigned& sequence_length = 3, const unsigned& frame_step = 3, const unsigned& frame_stride = 1)
```
Comment (rrawther) [inline rocAL/include/meta_data/meta_data_reader.h:63]: why is this changed from tot?

File: rocAL/include/pipeline/tensor.h [hunk 3]
```diff
@@ -303,7 +303,7 @@ class Tensor : public rocalTensor {
     void* buffer() { return _mem_handle; }
     vx_tensor handle() { return _vx_handle; }
     vx_context context() { return _context; }
-    void set_mem_handle(void* buffer) { _mem_handle = buffer; }
+    void set_mem_handle(void* buffer) override { _mem_handle = buffer; }
```
Comment (rrawther) [inline rocAL/include/pipeline/tensor.h:306]: It is better to make this function bool and return false if the caller is sending an invalid mem_handle

File: rocAL/source/api/rocal_api_data_loaders.cpp [hunk 4]
```diff
@@ -302,7 +297,7 @@ rocalSequenceReader(
         output = context->master_graph->create_loader_output_tensor(info);
         auto cpu_num_threads = context->master_graph->calculate_cpu_num_threads(1);
 
-        context->master_graph->add_node<ImageLoaderNode>({}, {output})->init(internal_shard_count, cpu_num_threads, source_path, "", std::map<std::string, std::string>(), StorageType::SEQUENCE_FILE_SYSTEM, DecoderType::TURBO_JPEG, shuffle, loop, context->master_graph->sequence_batch_size(), context->master_graph->mem_type(), context->master_graph->meta_data_reader(), decoder_keep_original, "", sequence_length, step, stride);
+        context->master_graph->add_node<ImageLoaderNode>({}, {output})->init(internal_shard_count, cpu_num_threads, source_path, "", std::map<std::string, std::string>(), StorageType::SEQUENCE_FILE_SYSTEM, DecoderType::TURBO_JPEG, shuffle, loop, context->master_graph->sequence_batch_size(), context->master_graph->mem_type(), context->master_graph->meta_data_reader(), decoder_keep_original, "", sequence_length, step, stride, ExternalSourceFileMode::NONE);
```
Comment (rrawther) [inline rocAL/source/api/rocal_api_data_loaders.cpp:300]: if the last argument is defalt to ExternalSourceFileMode::NONE, no need to pass that in this call

File: rocAL/source/api/rocal_api_data_loaders.cpp [hunk 5]
```diff
@@ -2024,6 +2019,66 @@ rocalRawCIFAR10Source(
...
+
+        auto info = TensorInfo(std::move(dims),
+                               context->master_graph->mem_type(),
+                               RocalTensorDataType::UINT8,
+                               tensor_layout,
+                               color_format);
+        output = context->master_graph->create_loader_output_tensor(info);
+        context->master_graph->set_external_source_reader_flag();
+        
+        unsigned shard_count = 1;  // Hardcoding the shard count to 1 for now.
```
Comment (rrawther) [inline rocAL/source/api/rocal_api_data_loaders.cpp:2063]: Probably, we need to support multi-shards for multi-GPU training with external source. Suggest to add that as a separate PR

File: rocAL/source/api/rocal_api_data_transfer.cpp [hunk 6]
```diff
@@ -79,6 +79,34 @@ void
...
+RocalStatus ROCAL_API_CALL
+rocalExternalSourceFeedInput(
+    RocalContext p_context,
+    const std::vector<std::string>& input_images_names,
+    bool is_labels,
+    const std::vector<unsigned char*>& input_buffer,
+    const std::vector<ROIxywh>& roi_xywh,
+    unsigned int max_width,
+    unsigned int max_height,
+    int channels,
```
Comment (rrawther) [inline rocAL/source/api/rocal_api_data_transfer.cpp:91]: please change it to unsigned

**Review by rrawther — CHANGES_REQUESTED** [review]
please address the review comments and warnings for readme

File: rocAL/include/meta_data/meta_data_reader.h (hunk 2 above)
Comment (SundarRajan98) [reply rocAL/include/meta_data/meta_data_reader.h]: Reverted back to the original version

**Comment by SundarRajan98** [comment]
@rrawther The markdown and spelling warnings are shown for all tests and files in the rocAL repo. I have fixed the warnings for the external source PR files but it will still show the warnings for other files in the repo

File: rocAL_pybind/rocal_pybind.cpp [hunk 7]
```diff
@@ -112,6 +145,13 @@ std::unordered_map<int, std::string> rocalToPybindOutputDtype = {
...
 PYBIND11_MODULE(rocal_pybind, m) {
     m.doc() = "Python bindings for the C++ portions of ROCAL";
+    // Bind the C++ structure
+    py::class_<ROIxywh>(m, "ROIxywh")
+        .def(py::init<>())
+        .def_readwrite("x", &ROIxywh::x)
+        .def_readwrite("y", &ROIxywh::y)
+        .def_readwrite("w", &ROIxywh::w)
+        .def_readwrite("h", &ROIxywh::h);
     // rocal_api.h
```
Comment (rrawther) [inline rocAL_pybind/rocal_pybind.cpp:155]: please remove commented line

**Review by rrawther — COMMENTED** [review]
@kiritigowda : I see a lot of warnings related to .md files. Do we need to fix those. Also I have tagged you for the review comment. Please check

File: rocAL/include/pipeline/tensor.h (hunk 3 above)
Comment (SundarRajan98) [reply rocAL/include/pipeline/tensor.h]: I have added a check inside this function and added throw condition in case the buffer ptr passed is invalid

File: rocAL_pybind/rocal_pybind.cpp (hunk 7 above)
Comment (SundarRajan98) [reply rocAL_pybind/rocal_pybind.cpp:149]: The comments are added in the pybind to mention the APIs that are present in each rocal_api headers. The comments are made for all rocal_api headers

**Comment by hansely** [comment]
@SundarRajan28 Thanks, looks good to me.

**Review by rrawther — APPROVED** [review]

**Comment by rrawther** [comment]
@kiritigowda : Can we merge this PR? Is there anything missing?
